Public/Src/Utilities/KeyValueStore/KeyValueStoreAccessor.cs

```diff
@@ -575,30 +665,30 @@ public KeyValueStoreAccessor(KeyValueStoreAccessor accessor)
         {
             using (m_rwl.AcquireReadLock())
             {
-                try
-                {
-                    if (Disabled)
-                    {
-                        return DisposedOrDisabledFailure;
-                    }
-
-                    return use(m_store, state);
-                }
-                catch (RocksDbSharpException ex)
+                // All RocksDb usages are checked for exceptions:
+                //  - If any error happens inside RocksDb, we assume the store is permanently corrupted. All 
+                //    further calls will result in an error. The user-defined invalidation handler is called.
+                //  - If an error happens within the user-provided function, but is not RocksDb related, then we 
+                //    don't disable and throw.
+                // In all cases, the user-defined failure handler is called.
+                if (Disabled)
                 {
-                    return HandleException(ex);
+                    return DisposedOrDisabledFailure;
                 }
-                // The SEHException class handles SEH (structured exception handling) errors that are thrown from unmanaged code, 
-                // but that have not been mapped to another .NET Framework exception. The SEHException class also corresponds to the HRESULT E_FAIL (0x80004005).
-                catch (System.Runtime.InteropServices.SEHException ex)
+
+                try
                 {
-                    return HandleException(ex);
+                    return use(m_store, state);
                 }
-                // Provide an opportunity for caller to handle any unknown exception type, including unmanaged ones, then rethrow
                 catch (Exception ex)
                 {
-                    HandleException(ex);
-                    throw;
+                    var result = HandleException(ex, out var rethrow);

```

Comment:
One way to deal with this pattern is to give an option to the caller (that provides an error handling function) to "handle" the error.

For instance, this pattern is used by TaskScheduler.UnobservedTaskException. The client subscribes to this event and in the callback they can set e.Handled = true; to avoid error propagation.

We can consider using the same pattern here, because it fits the needs, imo.

Reply:
Another benefit of the proposed approach is simplicity: I just quickly looked at stack walking logic inside HandleExcpetion function and it looked very scary to me.

---
In reply to: [324777386](https://github.com/microsoft/BuildXL/pull/879#discussion_r324777386) [](ancestors = 324777386)